### PR TITLE
Centralize decoding + use `defer` way more often

### PR DESCRIPTION
**Pull request by designatednerd**
We haven't really fully been taking advantage of Swift 4 `Codable` toys in order to keep our networking code simple. This PR is a start on doing some consolidation and simplification around that. 

This PR: 
- Vastly increases our use of `defer` to make sure a completion block is always fired on the main thread
- Centralizes parsing of timestamps across various APIs
- Centralizes parsing incoming JSON with `Decodable` 
- Prepares for centralization of parsing outgoing JSON 
- Pulls a few items out into their own `Codable` files so that we can leverage JSON decoding more easily

Future enhancements **not** in this PR: 
- Centralize generation of headers, especially signed headers
- Use more `Encodable` objects for sending stuff to the server
- Centralize creation of JSON data from `Encodable` objects

### CODE REVIEW

**Comment by designatednerd** [comment]
@yuliaveres I'm ok with holding off on anyone reviewing this until after #843 is merged if you'd prefer. Let me know what you think.

**Comment by designatednerd** [comment]
Per @yuliaveres we'll get hers merged first

**Comment by yuliaveres** [comment]
@designatednerd  #843  got merged.

**Comment by designatednerd** [comment]
@yuliaveres thanks for the heads up! I'll pull it in and deal with the merge conflicts 😄

**Comment by designatednerd** [comment]
This is ready for review.

File: Toshi/APIClients/EthereumAPIClient.swift [hunk 1]
```diff
@@ -212,50 +222,42 @@ final class EthereumAPIClient {
 
         // If we are on debug (specified in "other swift flags") we will mock out the collectiblesince they are only on production
         #if DEBUG
-            collectiblesTeapot = MockTeapot(bundle: Bundle(for: EthereumAPIClient.self), mockFilename: "getACollectible")
+            if !(collectiblesTeapot is MockTeapot) {
```
Comment (yuliaveres) [inline Toshi/APIClients/EthereumAPIClient.swift]: 👍

File: Toshi/APIClients/RatingsClient.swift [hunk 2]
```diff
@@ -142,7 +69,7 @@ class RatingsClient: NSObject {
             let hashedPayload = cereal.sha3WithID(string: payloadString)
             let signature = "0x\(cereal.signWithID(message: "POST\n\(path)\n\(timestamp)\n\(hashedPayload)"))"
 
-            let fields: [String: String] = ["Token-ID-Address": cereal.address, "Token-Signature": signature, "Token-Timestamp": String(describing: timestamp)]
+            let fields: [String: String] = ["Token-ID-Address": cereal.address, "Token-Signature": signature, "Token-Timestamp": timestamp]
```
Comment (yuliaveres) [inline Toshi/APIClients/RatingsClient.swift]: could you please change this to `Token-ID-Address`, `Token-Signature` and `Token-Timestamp` throughout the code to `Toshi-...`. Token is deprecated and not removed yet.

File: Toshi/Extensions/Codable+Parsing.swift [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+extension Encodable {
+
+    func toJSONData(with encoder: JSONEncoder = JSONEncoder(),
+                    successCompletion: (Data) -> Void,
+                    errorCompletion: (Error) -> Void) {
+        do {
+            let encoded = try encoder.encode(self)
+            successCompletion(encoded)
+        } catch let error {
+            errorCompletion(error)
```
Comment (yuliaveres) [inline Toshi/Extensions/Codable+Parsing.swift]: maybe we would like to have `DLog("Parsing error: \(error)")` here as well ?

File: Toshi/Extensions/Codable+Parsing.swift [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        do {
+            let decoded = try decoder.decode(Self.self, from: data)
+            successCompletion(decoded)
+        } catch let error {
+            DLog("Parsing error: \(error)")
+            errorCompletion?(.invalidResponseJSON)
+        }
+    }
+
+    static func optionalFromJSONData(_ data: Data,
```
Comment (yuliaveres) [inline Toshi/Extensions/Codable+Parsing.swift]: do we use this ?

**Pull request merged**